codeface/R/cluster/graph_comparison.r

```diff
@@ -142,37 +173,134 @@ graph.comparison <- function(g.1, g.2) {
+  # Get the database ids and resolve them to names
   global.ids.1 <- graph.data.1$v.global.ids
   node.label.1 <- sapply(global.ids.1, function(id)
         query.person.name(con, id))
-  vertex.df.1 <- data.frame(name=global.ids.1, Id=node.label.1)
   global.ids.2 <- graph.data.2$v.global.ids
   node.label.2 <- sapply(global.ids.2, function(id)
         query.person.name(con, id))
-  vertex.df.2 <- data.frame(name=global.ids.2, Id=node.label.2)
 
-  ## Create igraph objects
-  g.1 <- graph.data.frame(edgelist.1, vertices=vertex.df.1, directed=TRUE)
-  g.2 <- graph.data.frame(edgelist.2, vertices=vertex.df.2, directed=TRUE)
+  # Create two dataframes to be merged by developer name
+  temp.vertex.df.1 <- data.frame(id1=global.ids.1, name=node.label.1)
+  temp.vertex.df.2 <- data.frame(id2=global.ids.2, name=node.label.2)
+
+  # Merge the graphs by developers and create a mapping to local ids
+  temp.merged <- merge (temp.vertex.df.1, temp.vertex.df.2, all=T, by="name")
+  merged.vertex <- data.frame(id=1:nrow(temp.merged), name=temp.merged$name, id1=temp.merged$id1, id2=temp.merged$id2)
+  map1 <- get.index.map1(merged.vertex)
+  map2 <- get.index.map2(merged.vertex)
+
+  # Use the mappings to create a local edge and vertex list for graph1.
+  local.ids.1 <- map.ids(global.ids.1, map1)
+  edgelist.local.1 <- data.frame(from=map.ids(edgelist.1$fromId, map1),
+                         to=map.ids(edgelist.1$toId, map1),
+                         weight=edgelist.1$weight)
+  vertex.df.1 <- data.frame(id1=local.ids.1, name=node.label.1)
+
+  # Use the mappings to create a local edge and vertex list for graph2.
+  local.ids.2 <- map.ids(global.ids.2, map2)
+  edgelist.local.2 <- data.frame(from=map.ids(edgelist.2$fromId, map2),
+                                 to=map.ids(edgelist.2$toId, map2),
+                                 weight=edgelist.2$weight)
+  vertex.df.2 <- data.frame(id2=local.ids.2, name=node.label.2)
+
+  directed <- T
+  ## create graph instances and run the graph comparison.
+
+  # A graph with nno weights and only the local vertices
+  unmerged.unweighted.g.1 <- graph.data.frame(
+    data.frame(from=edgelist.local.1$from, to=edgelist.local.1$to),
+    vertices=vertex.df.1, directed=directed)
+  unmerged.unweighted.g.2 <- graph.data.frame(
+    data.frame(from=edgelist.local.2$from, to=edgelist.local.2$to),
+    vertices=vertex.df.2, directed=directed)
+  V(unmerged.unweighted.g.1)$Id <- vertex.df.1$name
+  V(unmerged.unweighted.g.2)$Id <- vertex.df.2$name
+
+  # A graph with weights and all vertices
+  g.1 <- graph.data.frame(edgelist.local.1, vertices=merged.vertex, directed=directed)
+  g.2 <- graph.data.frame(edgelist.local.2, vertices=merged.vertex, directed=directed)
+  V(g.1)$Id <- merged.vertex$name
+  V(g.2)$Id <- merged.vertex$name
+
+  # vertex.diff describes how different the edges are, we now calculate how different the graphs are
+  vertex.diff <- graph.comparison(g.1, g.2, weighted = weighted, symmetric = symmetric)
```

Comment:
To clarify, I expected
```R
# A graph with weights and all vertices
g.1 <- graph.data.frame(edgelist.local.1, vertices=vertex.df.1, directed=directed)
g.2 <- graph.data.frame(edgelist.local.2, vertices=vertex.df.2, directed=directed)
V(g.1)$Id <- vertex.df.1$name
V(g.2)$Id <- vertex.df.2$name
vertex.diff <- graph.comparison(g.1, g.2, weighted = weighted, symmetric = symmetric)
```
to give the same `vertex.diff` vector (but with less ones) but that doesn't seem to be the case. I'll try to investigate this.